server/src/main/java/org/opensearch/cluster/metadata/QueryGroupMetadata.java

```diff
@@ -0,0 +1,226 @@
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.core.ParseField;
+import org.opensearch.core.common.Strings;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.ConstructingObjectParser;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.opensearch.cluster.metadata.Metadata.ALL_CONTEXTS;
+
+/**
+ * This class holds the QueryGroupMetadata
+ * sample schema
+ * {
+ *     "queryGroups": {
+ *         "_id": {
+ *             {@link QueryGroup}
+ *         },
+ *        ...
+ *     }
+ * }
+ */
+@ExperimentalApi
+public class QueryGroupMetadata implements Metadata.Custom {
+    public static final String TYPE = "queryGroup";
+    private static final ParseField QUERY_GROUP_FIELD = new ParseField("queryGroups");
+
+    @SuppressWarnings("unchecked")
+    static final ConstructingObjectParser<QueryGroupMetadata, Void> PARSER = new ConstructingObjectParser<>(
+        "queryGroupParser",
+        args -> new QueryGroupMetadata((Set<QueryGroup>) args[0])
+    );
+
+    static {
+        PARSER.declareObjectArray(
+            ConstructingObjectParser.constructorArg(),
+            (p, c) -> QueryGroup.fromXContent(p),
+            QUERY_GROUP_FIELD
+        );
+    }
+
+    private final Set<QueryGroup> queryGroups;
+
+    public QueryGroupMetadata(Set<QueryGroup> queryGroups) {
+        this.queryGroups = queryGroups;
+    }
+
+    public QueryGroupMetadata(StreamInput in) throws IOException {
+        this.queryGroups = in.readSet(QueryGroup::new);
+    }
+
+    public Set<QueryGroup> queryGroups() {
+        return this.queryGroups;
+    }
+
+    /**
+     * Returns the name of the writeable object
+     */
+    @Override
+    public String getWriteableName() {
+        return TYPE;
+    }
+
+    /**
+     * The minimal version of the recipient this object can be sent to
+     */
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.V_3_0_0;
+    }
+
+    /**
+     * Write this into the {@linkplain StreamOutput}.
+     *
+     * @param out
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeCollection(queryGroups);
+    }
+
+    /**
+     * @param builder
+     * @param params
+     * @return
+     * @throws IOException
+     */
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(QUERY_GROUP_FIELD.getPreferredName(), queryGroups);
+        builder.endObject();
+        return builder;
+    }
```

Comment:
We don't need the builder.startObject(); and builder.endObject(); We can delete those lines.
    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
        builder.field(QUERY_GROUP_FIELD.getPreferredName(), queryGroups);
        return builder;
    }

Reply:
Right! since the start and end object is present in `Metadata` class's toXContent method.